### DISCUSSION TITLE
Why Homebrew always build from sources after "brew install package"?

### DISCUSSION THREAD

**Discussion by Xiyuer2198** [Homebrew/discussions#162] [Everyday usage]
Here is the command output:

❯ brew install htop                   
Updating Homebrew...
Warning: You are using macOS 11.1.
We do not provide support for this pre-release version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
pre-release version.

==> Downloading https://ftp.gnu.org/gnu/autoconf/autoconf-2.69.tar.gz
######################################################################## 100.0%
==> Downloading https://git.savannah.gnu.org/cgit/config.git/snapshot/config-0b5188819ba6091770064adf26360b204113317e.tar.gz
 -#O#-  #   #                                                                 
==> Downloading https://ftp.gnu.org/gnu/automake/automake-1.16.2.tar.xz
######################################################################## 100.0%
==> Downloading https://github.com/Homebrew/formula-patches/raw/e5fbd46a25e35663059296833568667c7b572d9a/libtool/dynamic_lookup-11.patch
==> Downloading from https://raw.githubusercontent.com/Homebrew/formula-patches/e5fbd46a25e35663059296833568667c7b572d9a/libtool/dynamic_lookup-11.patch
######################################################################## 100.0%
==> Downloading https://ftp.gnu.org/gnu/libtool/libtool-2.4.6.tar.xz
######################################################################## 100.0%
==> Downloading https://pkgconfig.freedesktop.org/releases/pkg-config-0.29.2.tar.gz
######################################################################## 100.0%
==> Downloading https://ftp.gnu.org/gnu/ncurses/ncurses-6.2.tar.gz
######################################################################## 100.0%
==> Downloading https://github.com/htop-dev/htop/archive/3.0.2.tar.gz
==> Downloading from https://codeload.github.com/htop-dev/htop/tar.gz/3.0.2
######################################################################## 100.0%
==> Installing dependencies for htop: autoconf, automake, libtool, pkg-config and ncurses
==> Installing htop dependency: autoconf
==> ./configure --prefix=/usr/local/Cellar/autoconf/2.69 --with-lispdir=/usr/local/Cellar/autoconf/2.69/share/emacs/site-lisp/autoconf
==> make install
🍺  /usr/local/Cellar/autoconf/2.69: 68 files, 3.0MB, built in 7 seconds
==> Installing htop dependency: automake
==> ./configure --prefix=/usr/local/Cellar/automake/1.16.2_1

**Comment by Xiyuer2198** [comment]
Homebrew config:

❯ brew config 
HOMEBREW_VERSION: 2.5.11
ORIGIN: https://github.com/Homebrew/brew
HEAD: 07b3aa69f8dc5098437cf2562864558bae2d45aa
Last commit: 4 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 7cb22cbf15ea9a6e17a2b0f69e89dd881746ae0d
Core tap last commit: 16 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
all_proxy: socks5://127.0.0.1:6153
http_proxy: http://127.0.0.1:6152
https_proxy: http://127.0.0.1:6152
Homebrew Ruby: 2.6.3 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit haswell
Clang: 12.0 build 1200
Git: 2.24.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.64.1 => /usr/bin/curl
macOS: 11.1-x86_64
CLT: 12.3.0.0.1.1605054730
Xcode: N/A

**Reply by SMillerDev** [reply]
> Warning: You are using macOS 11.1.
We do not provide support for this pre-release version.
You will encounter build failures with some formulae.
Please create pull requests instead of asking for help on Homebrew's GitHub,
Twitter or any other official channels. You are responsible for resolving
any issues you experience while you are running this
pre-release version.

Here's probably your answer. You can check and fix the issues in `brew doctor`, but more likely it's caused by running beta software.

**Reply by Xiyuer2198** [reply]
Thanks for your response. It seems that only I can do is wait...

**Comment by afberendsen** [comment]
I have the same issue. I know that, on my case, I changed some brew configuration file to always force compile...and now I do not remember where I made this change :(

I had a problem with the version of MacOS and the hardware platform. Executing apps was generating error 13. After some investigations, it was caused by a new version of MacOS (10.15) on a non-MacOS 11.15 supported hardware. Mac OS 11.15 assumes a minimum CPU type, which was making homebrew to use a pre-compiled binary for that minimum CPU. The minimum CPU had some opcodes that my current CPU does not support. So, two options were possible (1) change homebrew to always compile (2) downgrade to MacOS 11.13. I tried option (1) and then option (2), forgetting to change the configuration file back to its original contents. Now homebrew always compile when installing and I do not remember where the configuration file is....

```
HG000008:local afberendsen$ brew --config
HOMEBREW_VERSION: 3.0.10
ORIGIN: https://github.com/Homebrew/brew
HEAD: 0ad2564c5f1ddf6f16f707f2861556b80546713c
Last commit: 2 days ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: c2dfb7abb5c80146ed72a9bde3d6aa72838b14b1
Core tap last commit: 46 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 4
Homebrew Ruby: 2.6.3 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.6.3_2/bin/ruby
CPU: quad-core 64-bit sandybridge
Clang: 10.0 build 1000
Git: 2.17.2 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.54.0 => /usr/bin/curl
macOS: 10.13.6-x86_64
CLT: 10.1.0.0.1.1539992718
Xcode: N/A
```

**Reply by SMillerDev** [reply]
What does `brew doctor` say?

**Reply by carlocab** [reply]
Does
```
brew install hello
```
also build from source? It should be pouring from a bottle. The reason why it seems like you're always building from source is that you've only been installing formulae that don't have bottles for High Sierra anymore.

**Reply by MarcPhilippeJoly** [reply]
> The reason why it seems like you're always building from source is that you've only been installing formulae that don't have bottles for High Sierra anymore.

So the question is: how do I do the job of the people who aren't making bottles for _High Sierra_ or _Big Sur_ anymore ? How do I put things in bottle myself, when having them as puddles on the floor is a mess?

**Reply by p-linnane** [reply]
You can build your own bottles if you want, but you still have to build from source: https://docs.brew.sh/Bottles#creation

**Reply by carlocab** [reply]
You can create a bottle using
```
brew install --build-bottle foo
brew bottle foo
```
and then install the bottle using
```
brew install /path/to/foo.bottle.tar.gz
```
If you'd like to distribute that bottle, that's a whole other story.

**Reply by MarcPhilippeJoly** [reply]
> You can create a bottle using
> 
> ```
> brew install --build-bottle foo
> brew bottle foo
> ```

I think you've missed the point that the main problem is that *brew is no longer usable for building things* on older (but perfectly functional and extremely more powerful than their users actually need) 2014 machines.
Brew downloads compilers when needed, but fails to build the compilers required to build the most interesting applications.
The idea is to *build a bottle for brew without using brew.* The link https://docs.brew.sh/Bottles#creation provided by [p-linnane](https://github.com/p-linnane) gives an idea.

It's mainly a *political obsolescence* issue and a *denial of backward compatibility* issue, more than a lack of documentation about Homebrew.

**Reply by p-linnane** [reply]
We don't support macOS older than Monterey, and that will be going away this month. Once that happens, we will no longer build Monterey bottles. Formula version updates will happen and bottles will only be built for Ventura, Sonoma, Sequoia, and Linux. As time goes on, there will be fewer and fewer Monterey bottles available. This is our standard practice every year when a new macOS comes out. We match Apple's supported versions.

If upgrading hardware is not an option, there are tools that will allow you to upgrade macOS beyond what Apple supports. That's not something we officially support, but it may work for your purposes.

**Reply by carlocab** [reply]
> I think you've missed the point

If I've missed a point, then that might be because the initial question was poorly stated:

> So the question is: how do I do the job of the people who aren't making bottles for _High Sierra_ or _Big Sur_ anymore ? How do I put things in bottle myself, when having them as puddles on the floor is a mess?

> that the main problem is that _brew is no longer usable for building things_ on older (but perfectly functional and extremely more powerful than their users actually need) 2014 machines. Brew downloads compilers when needed, but fails to build the compilers required to build the most interesting applications. 

If you're on a 2014 machine, then you will likely be able to install Catalina or Big Sur. On those macOS versions, you can, from a bottle, `brew install llvm@12`, which is a recent enough compiler to be able to build most things.

Some things that require a very recent C++ standard will fail to build, but you can probably bootstrap your way to a newer LLVM version using `llvm@12`. Recent versions of LLVM will likely fail to build entirely, though. (Probably 16 or newer.)

> It's mainly a _political obsolescence_ issue and a _denial of backward compatibility_ issue

I disagree. It's really more an issue of practicality. Our resources are limited, so we choose to focus what we have on supporting the three most recent versions of macOS. These are also the versions that [most of our users use](https://formulae.brew.sh/analytics/os-version/365d/).

This isn't political at all. As for backward compatibility, bottles built for newer OS versions just don't work on older ones. That's just not in our control. Perhaps you can [take it up with Apple](https://developer.apple.com/bug-reporting/).
